go/controller/internal/reconciler/reconciler.go

```diff
@@ -333,59 +333,69 @@ func (a *kagentReconciler) ReconcileKagentMCPServer(ctx context.Context, req ctr
 }
 
 func (a *kagentReconciler) ReconcileKagentRemoteMCPServer(ctx context.Context, req ctrl.Request) error {
-	// reconcile the agent team itself
-	toolServer := &v1alpha2.RemoteMCPServer{}
-	if err := a.kube.Get(ctx, req.NamespacedName, toolServer); err != nil {
-		// if the tool server is not found, we can ignore it
+	serverRef := req.NamespacedName.String()
+	l := reconcileLog.WithValues("remoteMCPServer", serverRef)
+
+	server := &v1alpha2.RemoteMCPServer{}
+	if err := a.kube.Get(ctx, req.NamespacedName, server); err != nil {
+		// if the remote MCP server is not found, we can ignore it
 		if k8s_errors.IsNotFound(err) {
 			// Delete from DB if the remote mcp server is deleted
 			dbServer := &database.ToolServer{
-				Name:      utils.GetObjectRef(toolServer),
-				GroupKind: schema.GroupKind{Group: "kagent.dev", Kind: "RemoteMCPServer"}.String(),
+				Name:      serverRef,
+				GroupKind: server.GroupVersionKind().GroupKind().String(),
 			}
+
 			if err := a.dbClient.DeleteToolServer(dbServer.Name, dbServer.GroupKind); err != nil {
-				reconcileLog.Error(err, "failed to delete tool server for remote mcp server", "remoteMCPServer", utils.GetObjectRef(toolServer))
+				l.Error(err, "failed to delete tool server for remote mcp server")
 			}
-			reconcileLog.Info("remote mcp server was deleted", "remoteMCPServer", utils.GetObjectRef(toolServer))
+
 			if err := a.dbClient.DeleteToolsForServer(dbServer.Name, dbServer.GroupKind); err != nil {
-				reconcileLog.Error(err, "failed to delete tools for remote mcp server", "remoteMCPServer", utils.GetObjectRef(toolServer))
+				l.Error(err, "failed to delete tools for remote mcp server")
 			}
+
 			return nil
 		}
-		return fmt.Errorf("failed to get tool server %s: %v", req.Name, err)
+
+		return fmt.Errorf("failed to get remote mcp server %s: %v", serverRef, err)
 	}
 
 	dbServer := &database.ToolServer{
-		Name:        utils.GetObjectRef(toolServer),
-		Description: toolServer.Spec.Description,
-		GroupKind:   schema.GroupKind{Group: "kagent.dev", Kind: "RemoteMCPServer"}.String(),
+		Name:        serverRef,
+		Description: server.Spec.Description,
+		GroupKind:   server.GroupVersionKind().GroupKind().String(),
+	}
+	tools, err := a.upsertToolServerForRemoteMCPServer(ctx, dbServer, &server.Spec)
+	if err != nil {
+		l.Error(err, "failed to upsert tool server for remote mcp server")
+
+		// Fetch previously discovered tools from database if possible
+		var discoveryErr error
+		tools, discoveryErr = a.getDiscoveredMCPTools(ctx, serverRef)
+		if discoveryErr != nil {
+			err = multierror.Append(err, discoveryErr)
+		}
```

Comment:
Lifted this out of the status reconciliation and used the opportunity to optimise things a bit - since we are already list all tools as part of the upsert, we only need the database call if the upsert fails.